Introduce a tier-1 JIT compiler based on x86-64 architecture

qwe661234 opened the pull request:
When the using frequency of a block exceeds a predetermined threshold, the tier-1 JIT compiler traces the chained block and generate  corresponding low quailty machine code. The resulting target machine code is stored in the code cache for future utilization.

The primary objective of introducing the tier-1 JIT compiler is to enhance the execution speed of RISC-V instructions. This implementation requires two additional components: a tier-1 machine code generator,  and code cache. Furthermore, this tier-1 JIT compiler serves as the foundational target for future improvements.

In addition, we have developed a Python script that effectively traces code templates and automatically generates JIT code templates. This approach eliminates the need for manually writing duplicated code.

As shown in the performance analysis below, the tier-1 JIT compiler's  performance closely parallels that of QEMU in benchmarks with a  constrained dynamic instruction count. However, for benchmarks  featuring a substantial dynamic instruction count or lacking specific hotspots—examples include pi and STRINGSORT—the tier-1 JIT compiler demonstrates noticeably slower execution compared to QEMU.

Hence, a robust tier-2 compiler is essential to generate optimized machine code across diverse execution paths, coupled with a runtime profiler for detecting hotspots.

* Performance

| Metric   | rv32emu(T1C) | qemu  |
|----------|-----------------|-------|
|aes	   |             0.02|  0.031|
|mandelbrot|	        0.029| 0.0115|
|puzzle	   |           0.0115|	0.009|
|pi	       |           0.0413| 0.0177|
|dhrystone |	        0.331|	0.393|
|Nqeueens  |	        0.854|	0.749|
|qsort-O2  |	        2.384|	 2.16|
|miniz-O2  |	         1.33|	 1.01|
|primes-O2 |	         2.93|	1.069|
|sha512-O2 |	        2.057|	0.939|
|stream	   |           12.747|	10.36|
|STRINGSORT|	       89.012| 11.496|


As demonstrated in the memory usage analysis below, the tier-1 JIT compiler utilizes less memory than QEMU across all benchmarks.

* Memory usage

| Metric   | rv32emu(T1C) |   qemu  |
|----------|-----------------|---------|
|aes	   |          186,228|1,343,012|
|mandelbrot|	      152,203|  841,841|
|puzzle	   |          153,423|  890,225|
|pi	       |          152,923|  879,957|
|dhrystone |	      154,466|  856,404|
|Nqeueens  |	      154,880|  858,618|
|qsort-O2  |	      155,091|  933,506|
|miniz-O2  |	      165,627|1,076,682|
|primes-O2 |	      150,540|  928,446|
|sha512-O2 |	      153,553|	978,177|
|stream	   |          165,911|  957,845|
|STRINGSORT|	      167,871|1,104,702|

Related: #238

jserv reviewed (changes requested):
Eliminate the ubpf specific code. In fact, you already construct a new JIT backend.

jserv commented on src/riscv_private.h:
```diff
@@ -116,11 +110,32 @@ struct riscv_internal {
     block_map_t block_map; /**< basic block map */
+#else
+    struct cache *block_cache;
+#ifdef MIR
```
Drop MIR specific code.

jserv commented on src/rv32_template.c:
```diff
@@ -4,42 +4,73 @@
  * result is sign-extended to 64 bits.
  */
-RVOP(lui, { rv->X[ir->rd] = ir->imm; })
+RVOP(lui, { rv->X[ir->rd] = ir->imm; }, {RVCODEGEN(lui, {
```
Instead of `RVCODEGEN`, use `X64` identifier instead. Later, we may integrate A64 backend as well.

jserv replied:
Can you avoid the repeated `lui` declaration?

qwe661234 replied:
> Can you avoid the repeated `lui` declaration?

Does you mean rewrite to `RVCODEGEN({})`?

jserv replied:
I expect the instruction such as "lui" appears once within the scope of `RVOP` instead of twice.

jserv commented on src/rv32_template.c:
```diff
@@ -4,42 +4,73 @@
-RVOP(lui, { rv->X[ir->rd] = ir->imm; })
+RVOP(lui, { rv->X[ir->rd] = ir->imm; }, {RVCODEGEN(lui, {
+         emit_load_imm(state, RAX, ir->imm);
+         emit_store(state, S32, RAX, platform_parameter_registers[0],
```
Replace `platform_parameter_registers[0]` with something shorter.

jserv commented on src/rv32_template.c:
```diff
@@ -4,42 +4,73 @@
+RVOP(lui, { rv->X[ir->rd] = ir->imm; }, {RVCODEGEN(lui, {
+         emit_load_imm(state, RAX, ir->imm);
+         emit_store(state, S32, RAX, platform_parameter_registers[0],
+                    offsetof(struct riscv_internal, X) + 4 * ir->rd);
```
I anticipate the presence of template-based code generation, which necessitates minimal descriptions of target instructions without the need for complete C source code.

qwe661234 replied:
Does you mean writing IR here, and this IR would be parsed and translatedin to the current C code by python script?

jserv replied:
Exactly. The template-based JIT code generation is mapped to the use of a custom domain-specific language (DSL), which is interpreted by our Python script. Having minimal descriptions of target instructions without the need for complete C source code makes sense for both maintainability and future machine code lifting. See [Rellume — Lift machine code to LLVM IR](https://github.com/aengelke/rellume),

jserv replied:
Meanwhile, you can include a brief documentation section at the beginning of `rv32_template.c` explaining the use of DSL during JIT code generation.

jserv commented on src/cache.c:
```diff
@@ -3,20 +3,19 @@
  * exceeds the THRESHOLD, the JIT compiler flow is triggered.
  */
-#define THRESHOLD 1000
+#define THRESHOLD 4096
```
Add `FIXME` to indicate the need of an effective (dynamic) profiler.

jserv commented on Makefile:
```diff
@@ -118,6 +118,22 @@ gdbstub-test: $(BIN)
+OBJS_EXT += ubpf_jit_x86_64.o 
+uname_p := $(shell uname -p)
+ifneq ($(uname_p), x86_64)
+$(error JIT mode only support for x86_64 target now.)
```
Improve the grammar.

jserv reviewed (changes requested):
After removing the outdated ubpf-jit code, rename `ubpf_jit_x86_64.[ch]` to `jit-x64.[ch]`, which serves as the infrastructure for the tier-1 JIT compiler.

jserv commented on src/jit_x64.c:
```diff
@@ -0,0 +1,745 @@
+    BPF_REG_8,
+    BPF_REG_9,
+    BPF_REG_10,
+    _BPF_REG_MAX,
```
Refine these names for identifiers.

jserv replied:
Replace all identifiers which contain BPF with the refined ones.

jserv replied:
Apply consistent naming scheme. That is,
```c
enum vm_reg {
    VM_REG0 = 0,
    VM_REG1,
    ...
    VM_REG10,
    N_VM_REGS,
};
```

jserv commented on src/rv32_template.c:
```diff
@@ -3,43 +3,75 @@
+
 /* Internal */
-RVOP(nop, { rv->X[rv_reg_zero] = 0; })
+RVOP(nop, { rv->X[rv_reg_zero] = 0; }, {X64({})})
```
How about using shorter form? That is,
```c
RVOP(nop, { rv->X[rv_reg_zero] = 0; }, X64 { } )
```

For interpreter path, we can define a dummy `X64` macro which ignores everything.

qwe661234 replied:
The compiler would print error for X64 {} because the MACRO of RVOP are extended incorrectly.

jserv replied:
We can still stick to the shorter form `RVOP(nop, { rv->X[rv_reg_zero] = 0; }, X64({ }) )`, right?

qwe661234 replied:
This form is OK.

jserv commented on Makefile:
```diff
@@ -118,6 +118,22 @@ gdbstub-test: $(BIN)
+ifeq ($(call has, JIT), 1)
+OBJS_EXT += jit_x64.o 
+uname_p := $(shell uname -p)
+ifneq ($(uname_p), x86_64)
```
Reuse the variable `processor` defined in `mk/toolchain.mk`.

jserv commented on Makefile:
```diff
@@ -118,6 +118,22 @@ gdbstub-test: $(BIN)
+$(call set-feature, JIT)
+ifeq ($(call has, JIT), 1)
+OBJS_EXT += jit_x64.o 
+uname_p := $(shell uname -p)
```
Drop unnecessary `uname -p`.

jserv reviewed (changes requested):
Reuse processor detection in build system.

jserv commented on src/common.h:
```diff
@@ -6,7 +6,6 @@
 #pragma once
 
 #include "feature.h"
-
```
Don't change this line.

jserv commented on src/jit_x64.c:
```diff
@@ -0,0 +1,745 @@
+#include "state.h"
+#include "utils.h"
+
+enum bpf_register {
```
Since this C source is intended for internal use, specifically for JIT code generation, the naming scheme can follow the principle of simplicity. Therefore, we can safely drop the prefix `bpf_` or `BPF_`.

jserv commented on src/jit_x64.h:
```diff
@@ -0,0 +1,436 @@
+ * Generic x86-64 code generation functions
+ */
+
+#ifndef JIT_X64_H
```
Use `#pragma once` instead.

jserv commented on src/feature.h:
```diff
@@ -52,5 +52,10 @@
 #define RV32_FEATURE_ARC 0
 #endif
 
+/* Just-in-time compiler */
```
Rephrase as "Experimental just-in-time compiler"

jserv commented on src/jit_x64.h:
```diff
@@ -0,0 +1,436 @@
+
+#include "riscv_private.h"
+
+#define RAX 0
```
Use `enum` instead of defines for these registers.

jserv commented on src/jit_x64.h:
```diff
@@ -0,0 +1,436 @@
+
+struct jit_state *init_state(size_t size);
+
+uint32_t translate_x86_64(riscv_t *rv, block_t *block);
```
Always refer to the architecture as "x64" instead of "x86-64" since we are unlikely to support 32-bit x86 for the JIT compiler. We have plans and commitments to support two targets: x64 and Arm64.

jserv commented on src/cache.c:
```diff
@@ -3,20 +3,21 @@
  * "LICENSE" for information on usage and redistribution of this file.
  */
 
+#include "cache.h"
```
Don't change the order of local header inclusion.

jserv commented on src/jit_x64.c:
```diff
@@ -0,0 +1,745 @@
+// Copyright (c) 2015 Big Switch Networks, Inc
```
Drop the original licensing declaration since we are reworking the ubpf implementation. We can mention the following:
> This JIT implementation has undergone extensive modifications, heavily relying on the ubpf_jit_x86_64.[ch] from ubpf. The original ubpf_jit_x86_64.[ch] file served as the foundation and source of inspiration for adapting and tailoring it specifically for this JIT implementation. Therefore, credit and sincere thanks are extended to ubpf for their invaluable work.

Check src/map.[ch] for the way in which we provide credit for the previous work.

jserv commented on src/jit_x64.h:
```diff
@@ -0,0 +1,436 @@
+// Copyright (c) 2015 Big Switch Networks, Inc
```
Ditto. Rework the licensing notice and credit information.

jserv commented on src/jit_x64.h:
```diff
@@ -0,0 +1,436 @@
+    } else if (size == S32) {
+        /* mov */
+        emit1(state, 0x8b);
+    } else if (size == S64) {
```
Shall we care about S64 case?

jserv commented on src/emulate.c:
```diff
@@ -1030,11 +924,33 @@ void rv_step(riscv_t *rv, int32_t cycles)
-
-        /* execute the block */
+#if RV32_HAS(JIT)
+        /* execute by tiered 1 JIT compiler */
```
The correct term is "tier 1 JIT compiler" and "tiered compilation."

jserv commented on src/jit_x64.c:
```diff
@@ -0,0 +1,745 @@
+#define EBPF_MODE_JSLT 0xc0
+#define EBPF_MODE_JSLE 0xd0
+
+#define EBPF_OP_JA (EBPF_CLS_JMP | EBPF_MODE_JA)
```
Drop unused macros. Check if the following macros are required or not.

jserv commented on src/rv32_template.c:
```diff
@@ -3,43 +3,85 @@
  * specification version 20191213.
  */
 
+/* Please check files gen-jit-template.py and jit_x64.[ch] to understand custom
```
Improve the documentation.

jserv reviewed (changes requested):
Improve the git commit messages to explicitly highlight the performance comparison between the rv32emu interpreter, QEMU, and the proposed tier-1 JIT compiler. This comparison will focus on execution time and resource utilization, with particular attention to memory consumption.

jserv commented on src/emulate.c:
```diff
@@ -292,9 +166,14 @@ static block_t *block_alloc(riscv_t *rv)
     return block;
 }
 
+#if !RV32_HAS(JIT)
```
Function `map_hash` is only used by interpreter.

jserv commented:
This implementation has been successfully tested on x86-based macOS 10.15.7. However, note that, in compliance with Apple's security policy, additional entitlements are necessary to permit the execution of JIT-compiled code. See https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_security_cs_allow-jit

Quoted from [JDK-8234930](https://bugs.openjdk.org/browse/JDK-8234930) "Use `MAP_JIT` when allocating pages for code cache on macOS":
> This patch is a strong prerequest for mac-aarch64 port. it was made slightly different to upstream patch so on platforms other than macos this patch equals to "nothing".

qwe661234 commented:
> Improve the git commit messages to explicitly highlight the performance comparison between the rv32emu interpreter, QEMU, and the proposed tier-1 JIT compiler. This comparison will focus on execution time and resource utilization, with particular attention to memory consumption.

Currently, we don't have effective memory examination tool to examinate the real memory usage. The memory consumption we can gain is the result without counting `mmap`.

jserv commented on src/riscv.c:
```diff
@@ -143,10 +155,15 @@ bool rv_enables_to_output_exit_code(riscv_t *rv)
+#if !RV32_HAS(JIT)
     block_map_destroy(rv);
+#else
+    cache_free(rv->block_cache, NULL);
```
The second parameter is set to `NULL`, but there is no check for callback in the implementation of `cache_free`, resulting in segmentation fault.
Send another pull request to fix prior to the merge of this pull request.

jserv commented:
> Currently, we don't have effective memory examination tool to examinate the real memory usage. The memory consumption we can gain is the result without counting `mmap`.

Simply show what you have in git commit messages where we tend to track the development progress.

jserv commented on src/jit_x64.c:
```diff
@@ -0,0 +1,594 @@
+ * Reference:
+ *   https://github.com/iovisor/ubpf/blob/main/vm/ubpf_jit_x86_64.c
+ */
+
```
Add compilation guard. That is,
```c
#if !defined(__x86_64__)
#error "This implementation is dedicated to x86-64."
#endif
```

See https://github.com/sysprog21/kvm-host/blob/master/src/arch/x86/vm.c as an example.

jserv commented on src/jit_x64.c:
```diff
@@ -0,0 +1,594 @@
+#define MAX_INSTS 65536
+
+#if !defined(_countof)
+#define _countof(array) (sizeof(array) / sizeof(array[0]))
```
Use macro `ARRAYS_SIZE` defined in `src/common.h` instead.

jserv commented on src/jit_x64.c:
```diff
@@ -0,0 +1,594 @@
+#endif
+
+#if RV32_HAS(EXT_M)
+static void muldivmod(struct jit_state *state,
```
Is this forward declaration required?

jserv commented on src/jit_x64.c:
```diff
@@ -0,0 +1,594 @@
+ */
+
+#if defined(_WIN32)
+static int nonvolatile_reg[] = {RBP, RBX, RDI, RSI, R13, R14, R15};
```
Add `const` qualifier when possible.

jserv commented on src/emulate.c:
```diff
@@ -978,6 +987,10 @@ static block_t *block_find_or_translate(riscv_t *rv)
 }
 
+#if RV32_HAS(JIT)
+typedef void (*exec_tired1_block_func_t)(riscv_t *rv, uintptr_t);
```
Replace `exec_tired1_block_func_t ` with `exec_block_func_t ` because we don't have to distinguish from tier-1 and tier-2 compilers at the moment.

jserv replied:
In addition, "tired1" is considered as a typo.

jserv commented on src/emulate.c:
```diff
@@ -370,8 +382,14 @@ static bool is_branch_taken = false;
 static uint32_t last_pc = 0;
 
+#if RV32_HAS(JIT)
+/* record whether the block is replaced by cache. If so, clear the EBB
```
Refine the use of "EBB" since we move to block chaining.

jserv reviewed (changes requested):
Update CI pipeline to validate the JIT compiler.

jserv commented:
Consistently utilize the phrase "tier-1 JIT compiler" or "tier-1 compiler," and you can opt for the shortened form "T1C" for brevity.

jserv commented on src/emulate.c:
```diff
@@ -1030,11 +1051,33 @@ void rv_step(riscv_t *rv, int32_t cycles)
+        } /* check if using frequency of block exceed threshold */
+        else if ((block->backward &&
+                  cache_freq(rv->block_cache, block->pc_start) >= 1024) ||
+                 (cache_hot(rv->block_cache, block->pc_start))) {
```
Remove the parentheses out of `cache_hot`.

jserv commented on src/jit_x64.c:
```diff
@@ -0,0 +1,593 @@
+    VM_REG_8,
+    VM_REG_9,
+    VM_REG_10,
+    _VM_REG_MAX,
```
Replace `_VM_REG_MAX` with `N_VM_REGS` for consistent naming scheme.

jserv commented on src/jit_x64.c:
```diff
@@ -0,0 +1,593 @@
+#define X64_OP_MOD_REG (X64_CLS_ALU | X64_SRC_REG | 0x90)
+
+#define STACK_SIZE 512
+#define MAX_INSTS 1024
```
Replace `MAX_INSTS` with `MAX_INSNS` for consistency.

jserv commented on src/jit_x64.c:
```diff
@@ -0,0 +1,593 @@
+#undef _
+};
+/* clang-format on */
+typedef void (*codegen_tired1_block_func_t)(struct jit_state *,
```
Replace `codegen_tired1_block_func_t` with `codegen_block_func_t`.

jserv commented on src/rv32_template.c:
```diff
@@ -3,43 +3,86 @@
  * specification version 20191213.
  */
 
+/* Please review the files gen-jit-template.py and jit_x64.[ch] to comprehend
```
Rather than specifying "Please review the files gen-jit-template.py and jit_x64.[ch]," it is more beneficial to provide an overview of the domain-specific language (DSL) specification for JIT code generation. Documentation serves the purpose of simplifying maintenance tasks, rather than merely serving as an overly simplified reference.

jserv commented:
The proposed JIT compiler does not encompass effective code generation for RV32F. You should clarify which instruction sets and extensions are supported.

While launching the Quake via `make quake`, I am aware of the following:
```
Playing registered version.
Console initialized.
Exe: 20:12:56 Sep  1 2023
 8.0 megabyte heap
792k surface cache
Assertion failed: (NULL), function do_fmvwx, file src/rv32_jit_template.c, line 556.
```

Another reproducer:
```
 $ build/rv32emu build/maj2random.elf
maj_random (NROUNDS=2)                 count     mean(x)     mean(y) variance(x) variance(y)  std-dev(x)  std-dev(y)
(             0, (0 - 1K)/8K )    999999,057     0.49915     0.50594     0.08122     0.08092     0.28499     0.28447
Assertion failed: (NULL), function do_fmvwx, file src/rv32_jit_template.c, line 556.
```

jserv commented:
Tested on Ubuntu 20.04, Python 3.8 generates the following error messages:
```
  File "tools/gen-jit-template.py", line 153
    match items[0]:
          ^
SyntaxError: invalid syntax
make: *** [Makefile:130: src/rv32_jit_template.c] Error 1
$ python3 --version
Python 3.8.10
```

I installed Python 3.9.5 for comparison, and the above message still appear.
According to [Status of Python versions](https://devguide.python.org/versions/), Python 3.8 and 3.9 are still supported by Python Software Foundation.

qwe661234 commented:
> Tested on Ubuntu 20.04, Python 3.8 generates the following error messages:
> 
> ```
>   File "tools/gen-jit-template.py", line 153
>     match items[0]:
>           ^
> SyntaxError: invalid syntax
> make: *** [Makefile:130: src/rv32_jit_template.c] Error 1
> $ python3 --version
> Python 3.8.10
> ```
> 
> I installed Python 3.9.5 for comparison, and the above message still appear. According to [Status of Python versions](https://devguide.python.org/versions/), Python 3.8 and 3.9 are still supported by Python Software Foundation.

As alternative, I rewrite python switch into if-else.

jserv commented:
The JIT-compiled code can initiate Doom, but it crashes when attempting to start the game.

Tested on Ubuntu Linux 20.04.

jserv commented on src/rv32_template.c:
```diff
@@ -3,43 +3,123 @@
+ * RV32-A and RV32-F extension are unsupported.
+ */
+
+/* size: size of data
```
Add general descriptions prior to the following. Make them informative.

qwe661234 committed:
Introduce a tier-1 JIT compiler based on x86-64 architecture

When the using frequency of a block exceeds a predetermined threshold,
the tier-1 JIT compiler traces the chained block and generate
corresponding low quailty machine code. The resulting target machine
code is stored in the code cache for future utilization.

The primary objective of introducing the tier-1 JIT compiler is to
enhance the execution speed of RISC-V instructions. This implementation
requires two additional components: a tier-1 machine code generator,
and code cache. Furthermore, this tier-1 JIT compiler serves as the
foundational target for future improvements.

In addition, we have developed a Python script that effectively traces
code templates and automatically generates JIT code templates. This
approach eliminates the need for manually writing duplicated code.

As shown in the performance analysis below, the tier-1 JIT compiler's
performance closely parallels that of QEMU in benchmarks with a
constrained dynamic instruction count. However, for benchmarks
featuring a substantial dynamic instruction count or lacking specific
hotspots—examples include pi and STRINGSORT—the tier-1 JIT compiler
demonstrates noticeably slower execution compared to QEMU.

Hence, a robust tier-2 JIT compiler is essential to generate optimized
machine code across diverse execution paths, coupled with a runtime
profiler for detecting hotspots.

* Perfromance
| Metric   | rv32emu-T1C | qemu  |
|----------+-------------+-------|
|aes	   |         0.02|  0.031|
|mandelbrot|	    0.029| 0.0115|
|puzzle	   |       0.0115|  0.009|
|pi        |       0.0413| 0.0177|
|dhrystone |	    0.331|  0.393|
|Nqeueens  |	    0.854|  0.749|
|qsort-O2  |	    2.384|   2.16|
|miniz-O2  |	     1.33|   1.01|
|primes-O2 |	     2.93|  1.069|
|sha512-O2 |	    2.057|  0.939|
|stream	   |       12.747|  10.36|
|STRINGSORT|       89.012| 11.496|

As demonstrated in the memory usage analysis below, the tier-1 JIT
compiler utilizes less memory than QEMU across all benchmarks.

* Memory usage
| Metric   | rv32emu-T1C |   qemu  |
|----------+-------------+---------|
|aes	   |      186,228|1,343,012|
|mandelbrot|	  152,203|  841,841|
|puzzle	   |      153,423|  890,225|
|pi        |      152,923|  879,957|
|dhrystone |	  154,466|  856,404|
|Nqeueens  |	  154,880|  858,618|
|qsort-O2  |	  155,091|  933,506|
|miniz-O2  |	  165,627|1,076,682|
|primes-O2 |	  150,540|  928,446|
|sha512-O2 |	  153,553|  978,177|
|stream	   |      165,911|  957,845|
|STRINGSORT|      167,871|1,104,702|

Related: #238

jserv commented:
While T1C is capable of running Quake, it does not behave as expected when compared to the interpreter. I encountered an unexpected crash with Quake. To reproduce this, first run `make quake` and watch the demo scene without any keystrokes. You will then encounter a segmentation fault.
```
You got the shells
You got the Grenade Launcher
You receive 25 health
You get 2 rockets  
You got the rockets
You got the nailgun
You got the nails
You got the nails
You receive 15 health
You receive 15 health
You got the rockets
/bin/sh: line 1: 71739 Segmentation fault: 11  ../build/rv32emu quake.elf
```

Th log generated by the interpreter:
```
You got the shells
You got the Grenade Launcher
You receive 25 health
You get 2 rockets  
You got the rockets
You got the nailgun
You got the nails
You got the nails
You receive 15 health
You receive 15 health
You got the rockets
You get 2 rockets  
You got the nails
You got the nails
You receive 25 health
You got the gold key
Playing demo from demo2.dem.
You got the shells
You got the shells
You got the nails
You got armor
You got the nails
You receive 25 health
You got the Super Nailgun
You got the nails
You get 2 rockets  
```

jserv commented:
> Because quake has RV32F instructions, but T1C does not support it now.

We do need an effective fallback for RV32F and RV32A when the JIT compiler is unable to generate the corresponding machine code. A trampoline would be required for switching between translated blocks and non-JIT parts.
See https://github.com/jserv/rv32jit/blob/master/src/codegen/jitabi.cpp

jserv commented on src/decode.h:
```diff
@@ -32,150 +32,150 @@ enum op_field {
-        _(cjalr, 1, 2, ENC(rs1, rs2, rd))           \
-        _(cadd, 0, 2, ENC(rs1, rs2, rd))            \
-        _(cswsp, 0, 2, ENC(rs2))                    \
+#define RV_INSN_LIST                                   \
```
Refine the above comment for new field.

qwe661234 committed:
Avoid the tier-1 JIT compiler from translating RV32F and RV32A

jserv commented:
Finally, the JIT compiler is incorporated into this emulator repository.
